perf(transformer/refresh): use `take/take_in` instead of `drain`

Dunqing opened the pull request:
Calling `take` is cheaper than calling `drain`. These cases don't need to reuse the `Vec` after it drained, so we can replace them with `take`, which gives a tiny performance improvement.

<img width="633" alt="image" src="https://github.com/user-attachments/assets/a2a65957-09a0-47a5-9b9f-eaec576220c1" />

Dunqing commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10656?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#10656** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10656?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10656?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#10654** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10654?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=10656).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel reviewed (approved):
Good spot. Though I'm not actually sure how much `take_in` gains over `drain(..)` here. The code for `drain` contains more instructions, but in the case of `drain(..)` compiler may be able to optimize most of them out. *Maybe*. That's just guesswork.

I suspect the perf gain in this PR is more due to:

```diff
- let mut new_statements = ctx.ast.vec_with_capacity(program.body.len());
+ let mut new_statements = ctx.ast.vec_with_capacity(program.body.len() * 2);
```

...so `new_statements` will never need to grow and reallocate during the loop.

I can see another couple of things. I'll do a follow-up PR.

overlookmotel commented:
### Merge activity

* **Apr 29, 12:29 PM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

Dunqing committed:
perf(transformer/refresh): use `take/take_in` instead of `drain` (#10656)

Calling `take` is cheaper than calling `drain`. These cases don't need to reuse the `Vec` after it drained, so we can replace them with `take`, which gives a tiny performance improvement.

<img width="633" alt="image" src="https://github.com/user-attachments/assets/a2a65957-09a0-47a5-9b9f-eaec576220c1" />
